kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,422 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! The i8042 mouse driver.
+
+use alloc::{string::String, sync::Arc, vec::Vec};
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use aster_input::{
+    event_type_codes::{KeyCode, KeyStatus, RelCode, SynEvent},
+    input_dev::{InputCapability, InputDevice, InputEvent, InputId, RegisteredInputDevice},
+};
+use ostd::{
+    arch::{
+        irq::{MappedIrqLine, IRQ_CHIP},
+        trap::TrapFrame,
+    },
+    irq::IrqLine,
+};
+use spin::Once;
+
+use super::controller::{
+    I8042Controller, I8042ControllerError, I8042_CONTROLLER, PS2_ACK, PS2_BAT_OK, PS2_CMD_RESET,
+};
+use crate::alloc::string::ToString;
+
+const PACKET_LEN_STANDARD: usize = 3;
+const PACKET_LEN_INTELLIMOUSE: usize = 4;
+const PACKET_LEN_MAX: usize = 4;
+
+/// IRQ line for i8042 mouse.
+static IRQ_LINE: Once<MappedIrqLine> = Once::new();
+
+/// Registered device instance for event submission.
+static REGISTERED_DEVICE: Once<RegisteredInputDevice> = Once::new();
+
+/// ISA interrupt number for i8042 mouse.
+const ISA_INTR_NUM: u8 = 12;
+
+/// Mouse packet state machine.
+static MOUSE_PACKET_STATE: spin::Mutex<MousePacketState> =
+    spin::Mutex::new(MousePacketState::new());
+
+pub(super) fn init(controller: &mut I8042Controller) -> Result<(), I8042ControllerError> {
+    const DEVICE_ID_STANDARD: u8 = 0x00;
+    const DEVICE_ID_INTELLIMOUSE: u8 = 0x03;
+    const DEVICE_ID_INTELLIMOUSE_EXPLORER: u8 = 0x04;
+    const MOUSE_CMD_ENABLE_DATA_REPORTING: u8 = 0xF4;
+
+    // Reset mouse device by sending `PS2_CMD_RESET` to the second PS/2 port.
+    send_mouse_command(controller, PS2_CMD_RESET)?;
+
+    // The response should be `PS2_ACK` and `PS2_BAT_OK`, followed by the device PS/2 ID.
+    if controller.wait_and_recv_data()? != PS2_ACK {
+        return Err(I8042ControllerError::DeviceResetFailed);
+    }
+    // The reset command may take some time to finish. Try again a few times.
+    if (0..5).find_map(|_| controller.wait_and_recv_data().ok()) != Some(PS2_BAT_OK) {
+        return Err(I8042ControllerError::DeviceResetFailed);
+    }
+
+    let mut device_id = controller
+        .wait_and_recv_data()
+        .unwrap_or(DEVICE_ID_STANDARD);
```

Comment:
Why should we use `unwrap_or` here? If the device ID is not recognized, I think we should return `DeviceUnknown` like the keyboard case.
https://github.com/asterinas/asterinas/blob/ff021e0b89811c92069950a5a96ef41f9860de0b/kernel/comps/keyboard/src/i8042_chip/keyboard.rs#L52